feat(deps)!: upgrade odm layer to MongDB Rust driver v3.0.0

douggynix opened the pull request:
This pull request is related to that issue below I have opened to switch to the newer version of Mongodb Rust Driver **v3.0.0**
Issue : https://github.com/Devolutions/mongodm-rs/issues/29

Mongodb team introduced a lot of broken changes as a regard to  the former api in 2.x. Some methods have their arguments reduced, or some errors have been renamed. A lot of refactoring resulted in having build errors for features that have been also removed in Cargo.

This pull request is opened to address those. 
I have run the Unit test and it ran successfully. Feel free to review, fix, accept or deny this PR. 

Here is one of the PR from Mongodb Driver Rust team from which originates those broken changes from the APIs: https://github.com/mongodb/mongo-rust-driver/pull/1046
**FindOptions, FindOneOptions** and so on  are removed from find method arguments and  encapsulated by chain calling explicitly a with_options()
```rust
db.collection.find_one(doc).with_options(find_one_options) ;
db.collection.insert_one(&doc).with_options(insert_options).session(session);
```

![image](https://github.com/Devolutions/mongodm-rs/assets/11457752/e0abff38-8ada-431a-a5e4-f762a5318baf)

Additional PR from mongodb rust team with broken changes from the former 2.x api: //See driver pull request change for errors: https://github.com/mongodb/mongo-rust-driver/pull/1102

douggynix committed:
upgrade odm layer to mongodb rust driver v3.0.0

douggynix committed:
clean code

douggynix committed:
bump version to 1.0.0 for mongodb broken changes

CBenoit commented on Cargo.toml:
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "mongodm"
-version = "0.9.2"
+version = "1.0.0"
```
suggestion: Since we are still pre 1.0.0, it's fine to bump the minor number only. Cargo is treating those as breaking.

```suggestion
version = "0.10.0"
```

douggynix replied:
changes done by using the minor version as requested

douggynix committed:
bump the minor version number only

Since we are still pre 1.0.0, it's fine to bump the minor number only. Cargo is treating those as breaking.

Co-authored-by: Benoît Cortier <bcortier@proton.me>

douggynix committed:
revert back selection criteria with options

douggynix committed:
remove duplicate prefix markers

douggynix commented:
@CBenoit : I am done with my changes. Feel free to do the last reviews for refactoring if needed in order to have it merged.
I have also tested my changes as an external depency against a code based i am working on which depends on this crate with the mongodb rust driver v3.0.0. and it works

douggynix commented:
Any update or another review to have this merged?

CBenoit commented on Cargo.toml:
```diff
@@ -12,19 +12,20 @@ readme = "README.md"
+#default = ["tokio-runtime"]
+#tokio-runtime = ["mongodb/tokio-runtime"]
+#async-std-runtime = ["mongodb/async-std-runtime"]
+#chrono-0_4 = ["mongodb/bson-chrono-0_4"]
```
suggestion: Can you remove this completely if it’s not required anymore? I prefer deleting the lines instead of keeping them commented. Could you do that at the other places too?

douggynix replied:
I removed those commented lines  completely from `Cargo.toml`

CBenoit commented on src/index.rs:
```diff
@@ -369,7 +370,11 @@ pub async fn sync_indexes<CollConf: CollectionConfig>(
             let mut to_drop = Vec::new();
             for (i, index) in indexes.0.clone().into_iter().enumerate() {
                 let mut text_index_keys = None;
-                let index_doc = if index.keys.iter().any(|ind| matches!(ind, IndexKey::TextIndex(_))) {
+                let index_doc = if index
+                    .keys
+                    .iter()
+                    .any(|ind| matches!(ind, IndexKey::TextIndex(_)))
+                {
```
question: Do you know why the formatting was changed here? Is it the result of `cargo fmt` using the stable channel? I see other similar places.

douggynix replied:
The formatting has been changed due to cargo. 
I just ran cargo fmt as simple as that and it reformats it that way automatically. 
```shell
cargo fmt --verbose 
```
I am using rust tool chains version 1.79.0

CBenoit replied:
Sounds good. I just realized that our CI was dead and we were not checking the formatting anymore.

douggynix replied:
I noticed the CI was dead as well. We can address it from another PR or this one?

CBenoit replied:
I opened a PR for that already: https://github.com/Devolutions/mongodm-rs/pull/31

CBenoit commented on src/lib.rs:
```diff
@@ -178,13 +178,14 @@ pub mod prelude {
         JavaScriptCodeWithScope as BsonJavaScriptCodeWithScope, Regex as BsonRegex,
         Serializer as BsonSerializer, Timestamp as BsonTimestamp,
     };
+    //See driver pull request change for errors: https://github.com/mongodb/mongo-rust-driver/pull/1102
```
question: What is the purpose of this comment?

douggynix replied:
This is where i was able to trace back why WriteBulkFailure enum didn't exist and has been replaced in the driver from the pull request. There are a lot of changes from the the driver. i can remove it and put it in the PR description

CBenoit replied:
Yes, this comment is only useful where comparing the diff, but we don’t need it when looking at the new code

douggynix replied:
I remove it and reference it from the PR description if anyone or me ever needs it for tracking those changes.

CBenoit commented on src/lib.rs:
```diff
@@ -128,29 +128,29 @@ pub trait CollectionConfig {
 pub trait ToRepository {
     /// Shorthand for `Repository::<Model>::new`.
-    fn repository<M: Model>(&self) -> Repository<M>;
+    fn repository<M: Model + Send + Sync>(&self) -> Repository<M>;
```
question: Do we really need `Send + Sync` bounds everywhere? Should we put these bounds directly on the `Model` trait instead?

douggynix replied:
Good catch. I make **Model** trait **Send + Sync** and remove those bounds elsewhere and used it only where it is applicable.

CBenoit reviewed (commented):
Thank you! I went through it. I’m a bit surprised by the formatting, could you try to run `cargo fmt` again?

douggynix committed:
refactoring model with Send + Sync

CBenoit reviewed (approved):
This is looking good to me. I’ll merge and release a new version.
